enhancements/etcd/assisted-install.md

```diff
@@ -0,0 +1,223 @@
+1. The combination of all these things makes the situation unsupportable. 
+
+To allow toleration of this check we provided an unsupportedConfigOverrides
+`useUnsupportedUnsafeNonHANonProductionUnstableEtcd`. As assisted installer is
+now moving to GA a supported toleration is required.
+
+## Goals
+
+1.  Provide a supported solution for assisted installer that will allow
+    `cluster-etcd-operator` to tolerate 2 master replicas as install complete.
+
+## Non-Goals
+
+1.  Add support for less than 3 master node clusters
+1.  Support the current unsupportedConfigOverrides
+
+## Currently Supported Functionality 
+Currently `cluster-etcd-operator` and OpenShift only supports clusters with
+three master nodes as install complete.
+
+### EnvVarController
+Currently the `EnvVarController` ensures that all of the data required for the
+etcd static pods exists. A condition exists within the `getEtcdEnvVars`[1]  function which checks if we have the expected master node count or if
+`useUnsupportedUnsafeNonHANonProductionUnstableEtcd` is populated otherwise
+return error which results in the controller being degraded and no static-pod
+spec installed on the master node for etcd.
+
+[1] https://github.com/openshift/cluster-etcd-operator/blob/be2b8d97ca048baecb051d288971a9f6db411457/pkg/etcdenvvar/etcd_env.go#L71
+
+## Proposed changes
+The following changes are proposed.
+
+  - etcd-operator changes: TBD 
+  - API changes: https://github.com/openshift/installer/pull/4209
+
+In order for the operator to tolerate an unsafe install configuration all
+efforts must be taken to minimize the risk of tolerating user defined unsafe
+cluster. For example to save on costs user chooses 2 master replicas.
+
+1.  Observe installType from the install-config.
+1.  Observe master replica count from install-config as desired state.
+1.  Observe install status and only tolerate less than 3 masters during the time
+    of bootstrap complete until install complete. After install status is `complete`
+    the operator will go into a degraded state until the cluster has 3 or more master nodes.
+
+### cluster-config-v1 configmap
+The expected master replica count provided by the `install-config` is a very
+important value for this design. This value will be used as the desired state
+for nodes during install. By using this value we can protect against allowing a cluster which
+expects 3 master nodes to be install complete if one of those nodes fails to
+pull ignition for example.
+
+The installType provides clear intent of the installer. If the key exists with a
+value of `assisted` the cluster will tollerate less than 3 master replicas.
+
+cluster-config-v1 configmap with install-config embedded
+```
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: cluster-config-v1
+  namespace: kube-system
+data:
+  install-config: |
+    apiVersion: v1
+    baseDomain: group-b.devcluster.openshift.com
+    installType: assisted
```

Comment:
Rather than plumbing `installType` through openshift-install, could we have a separate ConfigMap that the assisted installer would add to the manifests and that the cluster-etcd-operator could read to determine whether to tolerate 2 master nodes?

Reply:
I am not against this approach if etcd is the only component that needs this information. It feels like this pattern of install augmentation was going to not be a one-off hence a more broad implementation but I guess we can revisit if it comes up again.

Reply:
My main objection with putting it in the install config is that it is not information upon which the installer will act in any way. As such, it is not install _config_. It is information that a user of the installer wants to pass into the cluster so that the cluster behaves in a unique way for that user. That fits pretty squarely into the purview of custom manifests.

Reply:
> I am not against this approach if etcd is the only component that needs this information. It feels like this pattern of install augmentation was going to not be a one-off hence a more broad implementation but I guess we can revisit if it comes up again.

Sure, that makes sense. I was expecting us to have several similar APIs for different components, but perhaps there's a way to express it through a shared API instead.

Reply:
> My main objection with putting it in the install config is that it is not information upon which the installer will act in any way. As such, it is not install _config_. It is information that a user of the installer wants to pass into the cluster so that the cluster behaves in a unique way for that user. That fits pretty squarely into the purview of custom manifests.

There are other examples of places where the installer uses a value to generate a manifest that it does not otherwise use itself. Isn't this the same sort of thing?

Reply:
It's fair that this is not the direct problem of the installer. The hope was that the intent of the install or the net strategy which would be used to actually deploy the cluster would still be governed by the installer thus part of the config. We are prepared to move on from this approach at this point.

Reply:
> > My main objection with putting it in the install config is that it is not information upon which the installer will act in any way. As such, it is not install _config_. It is information that a user of the installer wants to pass into the cluster so that the cluster behaves in a unique way for that user. That fits pretty squarely into the purview of custom manifests.
> 
> There are other examples of places where the installer uses a value to generate a manifest that it does not otherwise use itself. Isn't this the same sort of thing?

If the installer where creating a manifest based on the install config, then that could be acceptable field. In this case, the installer is not doing **anything** with the field.